### PR TITLE
E2e sign

### PR DESCRIPTION
**Pull request by loosebazooka**
@patflynn FYI

requires #53

### CODE REVIEW

**Comment by loosebazooka** [comment]
needs #56 before CI will pass

File: src/main/java/dev/sigstore/KeylessSigner.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.*;
+import java.security.cert.CertificateException;
+import java.security.spec.InvalidKeySpecException;
+import org.conscrypt.ct.SerializationException;
+
+public class KeylessSigner {
```
Comment (patflynn) [inline src/main/java/dev/sigstore/KeylessSigner.java:40]: javadoc with an example flow could be cool here.

File: src/main/java/dev/sigstore/KeylessSigner.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+      Signer signer) {
+    this.fulcioClient = fulcioClient;
+    this.fulcioVerifier = fulcioVerifier;
+    this.rekorClient = rekorClient;
+    this.rekorVerifier = rekorVerifier;
+    this.oidcClient = oidcClient;
+    this.signer = signer;
+  }
+
+  public static Builder builderForProd()
```
Comment (patflynn) [inline src/main/java/dev/sigstore/KeylessSigner.java:63]: I wonder if prod is the right wording here. Presumably users will be configuring this for either their own private sigstore infra, or one of the public providers.  Perhaps something more specific like sigstorePublicGood?  

I'm also wondering about usescases for doing something with this builder prior to building? In the tests it just doesBuilderForProd().build(), but why not cut out the middleperson?

File: src/main/java/dev/sigstore/KeylessSigner.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+      throws IOException, InvalidAlgorithmParameterException, CertificateException,
+          InvalidKeySpecException, NoSuchAlgorithmException {
+    var builder = new Builder();
+    var fulcioCert =
+        Resources.toByteArray(
+            Resources.getResource("dev/sigstore/tuf/production/fulcio_v1.crt.pem"));
```
Comment (patflynn) [inline src/main/java/dev/sigstore/KeylessSigner.java:69]: I still haven't got the TUF stuff straight in my head, so I'm curious how a key rotation would work here.

File: src/main/java/dev/sigstore/KeylessSigner.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        .fulcioClient(
+            FulcioClient.builder().build(),
+            FulcioVerifier.newFulcioVerifier(fulcioCert, ctfePublicKey))
+        .rekorClient(RekorClient.builder().build(), RekorVerifier.newRekorVerifier(rekorPublicKey))
+        .oidcClient(WebOidcClient.builder().build())
+        .signer(Signers.newEcdsaSigner());
+    return builder;
+  }
+
+  public static Builder builderForStaging()
```
Comment (patflynn) [inline src/main/java/dev/sigstore/KeylessSigner.java:84]: same naming concern here.

File: src/main/java/dev/sigstore/KeylessSigner.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    fulcioVerifier.verifyCertChain(signingCert);
+    // TODO: this signing workflow mandates SCTs, but fulcio itself doesn't, figure out a way to
+    // allow that to be known
+    fulcioVerifier.verifySct(signingCert);
+
+    // TODO: this reads the whole artifact which is probably expensive for large files, the signer
+    // interface should probably just sign non-digest artifacts instead of always applying sha256
+    // before signing. Then we can just apply the input stream to the sha256 calculator
+    // and send that digest to the signer
+    var artifactBytes = Files.readAllBytes(artifact);
```
Comment (patflynn) [inline src/main/java/dev/sigstore/KeylessSigner.java:177]: this should also be a stream shouldn't it?

File: src/main/java/dev/sigstore/KeylessSigner.java (hunk 3 above)
Comment (patflynn) [reply src/main/java/dev/sigstore/KeylessSigner.java:69]: aha I've discovered your readme below. perhaps a pointer from here?

File: src/main/java/dev/sigstore/KeylessSigner.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+      Preconditions.checkNotNull(rekorClient);
+      Preconditions.checkNotNull(rekorVerifier);
+      Preconditions.checkNotNull(oidcClient);
+      Preconditions.checkNotNull(signer);
+      return new KeylessSigner(
+          fulcioClient, fulcioVerifier, rekorClient, rekorVerifier, oidcClient, signer);
+    }
+  }
+
+  public void sign(Path artifact)
```
Comment (patflynn) [inline src/main/java/dev/sigstore/KeylessSigner.java:156]: should this return the rekor response?

File: src/main/java/dev/sigstore/KeylessSigner.java (hunk 5 above)
Comment (loosebazooka) [reply src/main/java/dev/sigstore/KeylessSigner.java:177]: What do you mean? IF you mean inputstream then yes, I need to actually do the right thing here. That's the whole TODO.

File: src/main/java/dev/sigstore/KeylessSigner.java (hunk 6 above)
Comment (loosebazooka) [reply src/main/java/dev/sigstore/KeylessSigner.java:156]: Should probably return a signature, a public key and a rekor response.

**Review by patflynn — APPROVED** [review]

**Pull request merged**